File: mnist_project/train.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

func loadCSV(path string, size int) []float64 {
	data := make([]float64, 0, size)

	file, err := os.Open(path)
	if err != nil {
		fmt.Printf("Error: Could not open file %s\n", path)
		return data
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 1024*1024*64), 1024*1024*64) 

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		for _, part := range parts {
			part = strings.TrimSpace(part)
			if part == "" {
				continue
			}
			val, err := strconv.ParseFloat(part, 64)
			if err == nil {
				data = append(data, val)
			}
			if len(data) >= size {
				return data
			}
		}
	}

	return data
}

func main() {
	fmt.Println("====================================")
	fmt.Println("        GO NATIVE AI BENCHMARK      ")
	fmt.Println("====================================\n")

	start := time.Now()

	trainX := loadCSV("train_X.csv", 40000)
	trainY := loadCSV("train_y.csv", 100)

	weights := make([]float64, 400)
	bias := 0.0
	lr := 0.1

	for epoch := 0; epoch < 20; epoch++ {
		for i := 0; i < 100; i++ {
			sum := bias
			for p := 0; p < 400; p++ {
				sum += trainX[i*400+p] * weights[p]
			}
			pred := 1.0 / (1.0 + math.Exp(-sum))
			err := trainY[i] - pred

			for p := 0; p < 400; p++ {
				weights[p] += lr * err * trainX[i*400+p]
			}
			bias += lr * err
		}
	}

	trainTime := time.Since(start)

	testX := loadCSV("test_X.csv", 4000)
	testY := loadCSV("test_y.csv", 10)

	correct := 0
	for i := 0; i < 10; i++ {
		sum := bias
		for p := 0; p < 400; p++ {
			sum += testX[i*400+p] * weights[p]
		}
		pred := 0.0
		if sum > 0.0 {
			pred = 1.0
		}
		if pred == testY[i] {
			correct++
		}
	}

	fmt.Println("--- GO RESULTS ---")
	fmt.Printf("Training Time:   %f seconds\n", trainTime.Seconds())
	fmt.Printf("Accuracy:        %d/10 correct\n", correct)
}

File: mnist_project/train.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <chrono>
#include <cmath>
#include <sstream>

using namespace std;
using namespace std::chrono;


vector<double> load_csv(string path, int size) {
    vector<double> data;
    ifstream file(path);
    string val;
    while (getline(file, val, ',')) {
        if (!val.empty()) data.push_back(stod(val));
        if (data.size() >= size) break;
    }
    return data;
}

int main() {
    auto start = high_resolution_clock::now();

    
    auto train_X = load_csv("train_X.csv", 40000);
    auto train_y = load_csv("train_y.csv", 100);
    
    vector<double> weights(400, 0.0);
    double bias = 0.0;
    double lr = 0.1;

    for (int epoch = 0; epoch < 20; epoch++) {
        for (int i = 0; i < 100; i++) {
            double sum = bias;
            for (int p = 0; p < 400; p++) {
                sum += train_X[i * 400 + p] * weights[p];
            }
            double pred = 1.0 / (1.0 + exp(-sum));
            double err = train_y[i] - pred;

            for (int p = 0; p < 400; p++) {
                weights[p] += lr * err * train_X[i * 400 + p];
            }
            bias += lr * err;
        }
    }

    auto end = high_resolution_clock::now();
    duration<double> time_span = duration_cast<duration<double>>(end - start);
    cout << "C++ Execution Time: " << time_span.count() << " seconds" << endl;
    return 0;
}